src/kernel/mpipeline.ml

```diff
@@ -155,19 +235,47 @@ let process
         | (`Interface _ | `Implementation _) as ast ->
           raw_source, Some ast
     )) in
-  let reader = timed_lazy reader_time (lazy (
-      let lazy source = source in
-      let config = Mconfig.normalize config in
-      Mocaml.setup_reader_config config;
-      let result = Mreader.parse ?for_completion config source in
-      result, config
+  let reader =
+    timed_lazy reader_time
+      (lazy
+        (let (lazy ((_, pp_result) as source)) = source in
+          let config = Mconfig.normalize config in
+          Mocaml.setup_reader_config config;
+          let { Reader_with_cache.output; cache_was_hit } =
+            let cache_disabling =
+              match (config.merlin.use_ppx_cache, pp_result) with
+              | false, _ -> Some "configuration"
+              | true, Some _ ->
+                (* The cache could be refined in the future to also act on the PP phase.
+                   For now, let's disable the whole cache when there's a PP. *)
+                Some "source preprocessor usage"
+              | true, None -> None
+            in
+            Reader_with_cache.apply ~cache_disabling ~force_invalidation:false
+              { source; for_completion; config }
```

Comment:
It just occur to me why this feels weird: the input of a `Reader_phase` can contain preprocessed AST and we ignore that in the fingerprinting.

I think a more natural way to disable the cache wen there is actually preprocessing would be to have the fingerprint be `None` (or some better name type constructor) if it is the case of `Some digest` if it is the raw source. Then `Fingerprint.equal` should return false if one of the two compared fingerprint if `None`.

That way we can remove the `cache_disabling` logic and simply pass `config.merlin.use_ppx_cache` or even `config` to the apply function.

Does that sounds reasonable ?

Reply:
I used to have the workflow you're proposing in the first commits (before the refactor). I changed it together with the more general refactor to what I have now, but that might have been a bad decision. Let me explain why I changed it. There are two things in which the two workflows differ from each other:

One is where the cache module captures the disabling of the cache. With the proposed/previous workflow, the disabling of the cache is captured in the fingerprint. With the current workflow, it's captured one level higher, i.e. in the cache itself (when the cache is disabled, then [the cache reference is set to `None`](https://github.com/ocaml/merlin/pull/1584/files#diff-7716196a07a66d255ac0cab613acff41943afd7c640a06b820a25d714954af99R32)). Capturing it at the level of the cache means that the cache neither stores a fingerprint nor a "last processing result" when being disabled. I think that makes sense since it's a bit weird to have a cache that stores a "last result" with a `None` fingerprint. What do you think? I'm not super familiar with how caches are usually implemented, so I might easily be missing something.

The other thing that's different is where the _logic of disabling_ the cache lives: with the current workflow, it lives outside the cache module and is passed to the `With_cache.apply` function from there. That has two upsides: 1. we stay more flexible in terms of disabling the cache; 2. it's visual from outside whether the cache is disabled or actually used. With the proposed/previous workflow, it is embedded in the cache logic: the cache is more of a black box from outside that's sometimes disabled and sometimes enabled, without knowing or being able to tweak it from outside which one of the two it is. The upside of that is that we don't need to pass the `cache_disabling` parameter. Btw, if you prefer the latter (which seems to be the case), I could change the workflow I'm talking about here even if I don't change the one from the last paragraph.

Reply:
Right, let's keep the more flexible version for now.